rest_api/controller/feedback.py

```diff
@@ -102,86 +63,107 @@ def eval_doc_qa_feedback(filters: FilterRequest = None):
+            text = document.text
+            # the final length of context(including the answer string) is 'context_size'.
+            # we try to add equal characters for context before and after the answer string.
+            # if either beginning or end of text is reached, we correspondingly
+            # append more context characters at the other end of answer string.
+            context_to_add = int((context_size - len(label.answer)) / 2)
+            start_pos = max(label.offset_start_in_doc - context_to_add, 0)
+            additional_context_at_end = max(context_to_add - label.offset_start_in_doc, 0)
+            end_pos = min(label.offset_start_in_doc + len(label.answer) + context_to_add, len(text) - 1)
+            additional_context_at_start = max(
+                label.offset_start_in_doc + len(label.answer) + context_to_add - len(text), 0
+            )
+            start_pos = max(0, start_pos - additional_context_at_start)
+            end_pos = min(len(text) - 1, end_pos + additional_context_at_end)
+            context = text[start_pos:end_pos]
+
+        if label.is_correct_answer is False and label.is_correct_document is False:  # No answer
+            squad_label = {
+                "paragraphs": [
+                    {
+                        "context": context,
+                        "id": label.document_id,
+                        "qas": [{"question": label.question, "id": label.id, "is_impossible": True, "answers": []}],
+                    }
+                ]
+            }
+        else:
+            squad_label = {
+                "paragraphs": [
+                    {
+                        "context": context,
+                        "id": label.document_id,
+                        "qas": [
+                            {
+                                "question": label.question,
+                                "id": label.id,
+                                "is_impossible": False,
+                                "answers": [{"text": label.answer, "answer_start": label.offset_start_in_doc}],
+                            }
+                        ],
+                    }
+                ]
+            }
+
+            # quality check
+            start = squad_label["paragraphs"][0]["qas"][0]["answers"][0]["answer_start"]
+            answer = squad_label["paragraphs"][0]["qas"][0]["answers"][0]["text"]
+            context = squad_label["paragraphs"][0]["context"]
+            if not context[start: start + len(answer)] == answer:
+                logger.error(
+                    f"Skipping invalid squad label as string via offsets "
+                    f"('{context[start:start + len(answer)]}') does not match answer string ('{answer}') "
+                )
+        export_data.append(squad_label)
 
     export = {"data": export_data}
 
+    with open("feedback_squad_direct.json", "w", encoding="utf8") as f:
+        json.dump(export_data, f, ensure_ascii=False, sort_keys=True, indent=4)
```

Comment:
API caller/client can sort the keys and add indent if they like, we can free server from spending time on sorting and reduce the payload size by not adding indent :)

Reply:
I guess it's a relatively small performance penalty for the added readability & convenience.